### ISSUE TITLE
Warnings / notifications make napari crash when called from a thread_worker

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#2681]
## 🐛 Bug

When using `warnings.warn()` from within a `thread_worker` napari freezes with this on stdout:
```
WARNING: QObject::setParent: Cannot set parent, new parent is in a different thread
```

I assume it freezes because the access to the GUI in main thread causes a crash.

## To Reproduce

```
import warnings

import napari
import time
from napari._qt.qthreading import thread_worker
import numpy as np

# create a viewer window
viewer = napari.Viewer()

# https://napari.org/guides/stable/threading.html
@thread_worker
def loop_run():
    while viewer.window.qt_viewer:  # loop until napari closes
        print("Hello world", time.time())
        warnings.warn("Time: " + str(time))
        time.sleep(0.5)

# Start the loop
worker = loop_run()
worker.start()

# Start napari
napari.run()
```

![image](https://user-images.githubusercontent.com/12660498/117564590-1c909400-b0ad-11eb-9732-9b6b82b419b2.png)

This code will run fine if `warnings.warn...` is commented out. 

## Expected behavior

I guess the notification manager could use some yield-mechanism to make the update to the GUI from the main thread.

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
napari: 0.4.8rc1
Platform: Windows-10-10.0.19041-SP0
Python: 3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 04:59:43) [MSC v.1916 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.2
SciPy: 1.6.2
Dask: 2021.04.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 - Build 27.20.100.8681
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- clEsperanto: 0.8.0
- console: 0.0.3
- napari-image-quality-analyzer: 0.1.0
- napari-skimage-regionprops: 0.0.0
- napari_webcam: 0.1.17
- scikit-image: 0.4.8rc1
- svg: 0.1.4
 - Any other relevant information:

**Comment by Czaki** [comment]
I think that it is classical problem. 
Qt widgets cannot be created outside the main thread. So trying to show warning created outside the main thread may produce such a problem. 
If I think correctly then there is a need to add a mechanism of piping warning to the main thread (for example using QT signal mechanism because Qt signals connected functions are executed inside the main thread).

**Comment by DragaDoncila** [comment]
This also happens when warnings occur in threads not managed by `thread_worker` e.g. `dask` threads, as @jni  and I discovered today.

**Comment by Czaki** [comment]
Because problem comes generally from using thread. 
EDIT

To call the method from the other thread the `invokeMethod` from `QMetaObject` could be used:
https://github.com/4DNucleome/PartSeg/blob/develop/package/PartSeg/common_backend/except_hook.py

**Comment by tlambert03** [comment]
Thanks @haesleinhuepf. relatively easy fix in #2688
